workspaces/diff-engine/src/events/http_interaction.rs

```diff
@@ -0,0 +1,399 @@
+use super::EventLoadingError;
+use avro_rs;
+use base64;
+use cqrs_core::Event;
+use serde::Deserialize;
+use serde_json;
+use std::io;
+
+// TODO: consider whether these aren't actually Events and the Traverser not an Aggregator
+
+#[derive(Deserialize, Debug)]
+pub struct HttpInteraction {
+  pub uuid: String,
+  pub request: Request,
+  pub response: Response,
+  pub tags: Vec<HttpInteractionTag>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct HttpInteractionTag {
+  name: String,
+  value: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Request {
+  pub host: String,
+  pub method: String,
+  pub path: String,
+  // #[serde(skip)]
+  pub query: ArbitraryData,
+  pub body: Body,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Response {
+  pub status_code: u16,
+  // #[serde(skip)]
+  pub headers: ArbitraryData,
+  pub body: Body,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Body {
+  pub content_type: Option<String>,
+  // #[serde(skip)]
+  pub value: ArbitraryData,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ArbitraryData {
+  shape_hash_v1_base64: Option<String>,
+  as_json_string: Option<String>,
+  as_text: Option<String>,
+}
+
+impl From<&ArbitraryData> for Option<serde_json::value::Value> {
+  fn from(data: &ArbitraryData) -> Option<serde_json::value::Value> {
+    if let Some(json_string) = &data.as_json_string {
+      Some(
+        serde_json::from_str(json_string)
+          .expect("as_json_string of ArbitraryData should always be valid json"),
+      )
+    } else if let Some(text) = &data.as_text {
+      Some(serde_json::Value::from(text.clone()))
+    } else if let Some(shape_hash) = &data.shape_hash_v1_base64 {
+      let decoded_hash = base64::decode(shape_hash)
+        .expect("shape_hash_v1_base64 of ArbitraryData should always be valid base64");
+      let json =
+        serde_json::from_slice(&decoded_hash).or_else::<ShapeHashParsingError, _>(move |err| {
+          let shape_hash_as_text = String::from_utf8(decoded_hash)?;
+          Ok(serde_json::Value::from(shape_hash_as_text))
+        });
+      match json {
+        Ok(json) => Some(json),
+        Err(json) => {
+          eprintln!("could not decode shape hash into a json value");
+          None
+        }
+      }
```

Comment:
I added this implementation for shape hashes going from my naïve idea of what they were, but now looking at what the shape hashes actually are, this implementation is obviously totally borked and non-sensical. Looking at the implementation of the shape hash, it's protobuffs, so trying to read that as json or a utf-8 string is going to fail miserably.